### PR TITLE
Pickup Items

### PR DESCRIPTION
**Pull request by joshuaskelly**
pickup-items
# Summary
Enables items to function an pickups. The basic behavior is to automatically pickup an item and put it in the player's inventory. Items can override this behavior and several do. For example, pickup up food eats it. Picking up a scroll casts it. Etc.

If a pickup is marked as floating, it will bob. Mesh items will also spin if marked as a pickup and floating.

### CODE REVIEW

File: Dungeoneer/src/com/interrupt/dungeoneer/entities/Item.java [hunk 1]
```diff
@@ -130,6 +142,8 @@
 	public transient String lastMeshFile = null;
 	public transient String lastTextureFile = null;
 
+	private Entity pickupHelper = null;
```
Comment (Interrupt) [inline Dungeoneer/src/com/interrupt/dungeoneer/entities/Item.java:145]: This could be marked 'private transient' to keep it from being serialized to disk

**Comment by joshuaskelly** [comment]
@Interrupt What should the behavior be if you pickup up weapon and then drop it from your inventory? Should it still be a pickup? Should it be a normal item?

**Comment by Interrupt** [comment]
@joshuaskelly it probably should still be a pickup, but maybe it should wait until the physics are asleep for the player to be able to pick it up again

**Comment by Interrupt** [comment]
@joshuaskelly I think the pickup logic just needs to check the `ignorePlayerCollision` variable when doing the pickup code, if set to true it means that the player just threw the object, and it will wait until it touches something / bounces off of a wall to flip that flag again. It's currently used to figure out when to damage the player when they throw an item.

**Comment by joshuaskelly** [comment]
I tried both of your suggestions and neither was a perfect fit.

- `ignorePlayerCollision` only gets reset if the item hits a wall. Hitting the floor does not reset that flag.
- `physicsSleeping` is always false for floating items.

**Comment by joshuaskelly** [comment]
I did make the pickup helper object a static transient object so we only have one and each pickup updates it right before it uses it.

**Review by Interrupt — APPROVED** [review]

**Comment by Interrupt** [comment]
Maybe items need to have a pickup cooldown timer on themselves, that gets set to some value when picked up?

**Pull request merged**